### ISSUE TITLE
doesn't apply right fix comment in jsx/tsx context

### ISSUE THREAD

**Issue by blackLearning** [microsoft/vscode-eslint#1074]
Can we check the context of code, when apply quick fix comment?

Expect:
![image](https://user-images.githubusercontent.com/15011394/93551632-f04ec880-f9a0-11ea-9061-a537fb1aab8e.png)

got:
![image](https://user-images.githubusercontent.com/15011394/93551531-acf45a00-f9a0-11ea-9db7-c9b6cf10b11c.png)

**Comment by dbaeumer** [comment]
See https://github.com/microsoft/vscode-eslint/pull/857. May be you want to extend that one to handle jsx/tsx context as well.

**Comment by dbaeumer** [comment]
I looked into this and it is actually hard to fix. ESLint is not giving me any context and it is therefore hard to decide whether I need to use `{ /* */}` or `//` to disable the rule

**Comment by robballou** [comment]
I kept running into this so I thought I would dig a little. (This is mostly thinking aloud here and maybe explaining for others that come looking...)

1. It seems like the issue is the `languageDefaults` [(src)](https://github.com/microsoft/vscode-eslint/blob/main/server/src/languageDefaults.ts#L16) will return `//` as the comment for `.jsx` files... which is sometimes fine. Used [here](https://github.com/microsoft/vscode-eslint/blob/main/server/src/eslintServer.ts#L661-L674)
2. From back in the TextMate days, you could get the scope of the cursor, but that doesn't exist in vscode 😞 
3. The only other thing I dug into was seeing if there is a way to say trigger the `editor.action.commentLine` command with text for the comment ... but I'm not seeing a clear way. My only theory right now is that you might be able to insert the disable text (ex `eslint-disable-next-line blah`), create a selection around that, then use the comment line action.

I feel like the add the text & execute the command may work out best since the language seems to know the correct comment type. No guarantees, but if I get time and figure out how to work on this locally I may try to get a PR going.

**Comment by rchl** [comment]
While this server is maybe a little vscode-centric ideally the solution would not be editor-specific as this is an LSP server used by many other editors also. Ideally the solution would come from the eslint.

**Comment by robballou** [comment]
That makes sense to me. I jumped in without that context so thank you for providing that. It also seems with things like #1499 it would also really complicate things on my "plan" 😀 

I also can't really find anything that looks promising on eslint's side to provide this information. They have [some scope information](https://eslint.org/docs/latest/developer-guide/working-with-rules#contextgetscope) but it seems more AST related than something that would provide something useful. I struggled with getting this extension running for debug (my own issue) so I couldn't really debug around what eslint is providing in runtime.

**Comment by MariaSolOs** [comment]
Hmmm so I looked into this a bit and maybe one could identify the embedded languages in the document and then modify the commenting accordingly. I wonder if this could be done by providing the languages configuration guide of the document to the VS Code extension, perhaps via the `Languages` namespace.

**Comment by MariaSolOs** [comment]
It would also be great if TSServer had a command for this. It already has one for auto-closing JSX tags, and so it already knows the context of the code at a specific line.

**Comment by rchl** [comment]
- again focusing on vscode
- tsserver is irrelevant here - it's not used by this server/extension

**Comment by MariaSolOs** [comment]
> * again focusing on vscode
> * tsserver is irrelevant here - it's not used by this server/extension

@rchl You're right, and trust me, I want this LSP server to remain useful to as many editors as possible. However, the solution I suggested (and this is just me throwing random ideas and I cannot guarantee them being feasible or anything) of using `Languages` would be done by the LSP client, and hence it can be specific to VS Code, since each editor will need to have its own client anyway.

And yes, TSServer isn't used in this extension, BUT the [built-in support for TypeScript](https://github.com/microsoft/vscode/tree/dbae720630e5996cc4d05c14649480a19b077d78/extensions/typescript-language-features/src/tsServer) does use it. My TSServer idea is more of an LSP proposal, of having a "commentRange" request that would let the editor figure out the right commenting. This idea comes from the fact that VS Code already has the logic for using the right comment tags in JSX, so I'm brainstorming on a way to leverage that.

**Comment by dbaeumer** [comment]
Looking into this in more detail it is actually pretty tricky. Delegating this back to the editor might not work in all cases either since the editor might not have smart language support installed for the language that ESLint is validating (consider there is no TS server available in VS Code because the extension got disabled; or the user simply didn't install smart language support for the given language). Relying on syntax tokens might not work either since even it TextMate such an embedded scope must not exist. 

IMO the right solution is still using ESLint itself to figure this out. Since I don't see ESLint adding such an information to the ESLint problems here is something I think should work:

- when linting a file we already ask ESLint to give us an expanded configuration object. This configuration object contains the ESLint parser used to parse the file.
- when we are in a file where we know the scope matters (tsx, jsx, ...) we could create an ESTree using the configured parser (https://github.com/estree/estree). The ESTree is standardized.
- we could then use the ESTree to figure out in which context we are

Given the following code sample:

```tsx
function App() {
  return (
    <div className= "App" >
    </div>
	);
}

export default App;
```

The corresponding ESTree looks like this:

<img width="571" alt="image" src="https://user-images.githubusercontent.com/1931590/183088811-642f11ff-c802-4ee4-b41f-04b14de46ac1.png">


We could then find the corresponding node based on the position and then walk the parent chain to see if we are in a `JSXElement`

This has the advantage that we don't use two different tools / parsers that might produce different information.

@MariaSolOs is you would like to look into this I am happy to help.
